stellargraph/mapper/knowledge_graph.py

```diff
@@ -0,0 +1,189 @@
+        target_ilocs = self.G._get_index_for_nodes(targets)
+
+        return KGTripleSequence(
+            max_node_iloc=self.G.number_of_nodes(),
+            source_ilocs=source_ilocs,
+            rel_ilocs=rel_ilocs,
+            target_ilocs=target_ilocs,
+            batch_size=self.batch_size,
+            shuffle=shuffle,
+            negative_samples=negative_samples,
+            seed=seed,
+        )
+
+
+class KGTripleSequence(Sequence):
+    def __init__(
+        self,
+        *,
+        max_node_iloc,
+        source_ilocs,
+        rel_ilocs,
+        target_ilocs,
+        batch_size,
+        shuffle,
+        negative_samples,
+        seed,
+    ):
+        self.max_node_iloc = max_node_iloc
+
+        num_edges = len(source_ilocs)
+        self.indices = np.arange(num_edges, dtype=np.min_scalar_type(num_edges))
+
+        self.source_ilocs = np.asarray(source_ilocs)
+        self.rel_ilocs = np.asarray(rel_ilocs)
+        self.target_ilocs = np.asarray(target_ilocs)
+
+        self.negative_samples = negative_samples
+
+        self.batch_size = batch_size
+        self.seed = seed
+
+        self.shuffle = shuffle
+
+        self._global_rs = np.random.RandomState(seed)
+        self._batch_samplers = []
+        self._global_lock = threading.Lock()
+
+    def _batch_sampler(self, batch_num):
+        self._global_lock.acquire()
+        try:
+            return self._batch_samplers[batch_num]
+        except IndexError:
+            new_samplers = batch_num - len(self._batch_samplers) + 1
+            seeds = self._global_rs.randint(2 ** 32, size=new_samplers)
+            self._batch_samplers.extend(np.random.RandomState(seed) for seed in seeds)
+            return self._batch_samplers[batch_num]
+        finally:
+            self._global_lock.release()
+
+    def __len__(self):
+        return int(np.ceil(len(self.indices) / self.batch_size))
+
+    def __getitem__(self, batch_num):
+        start = self.batch_size * batch_num
+        end = start + self.batch_size
+
+        indices = self.indices[start:end]
+
+        s_iloc = self.source_ilocs[indices]
+        r_iloc = self.rel_ilocs[indices]
+        o_iloc = self.target_ilocs[indices]
+        positive_count = len(s_iloc)
+        targets = None
+
+        if self.negative_samples is not None:
+            s_iloc = np.tile(s_iloc, 1 + self.negative_samples)
+            r_iloc = np.tile(r_iloc, 1 + self.negative_samples)
+            o_iloc = np.tile(o_iloc, 1 + self.negative_samples)
+
+            negative_count = self.negative_samples * positive_count
+            assert len(s_iloc) == positive_count + negative_count
+
+            rng = self._batch_sampler(batch_num)
+
+            change_source = rng.randint(2, size=negative_count).astype(bool)
+            source_changes = change_source.sum()
+            new_nodes = rng.randint(self.max_node_iloc, size=negative_count)
+
+            s_iloc[positive_count:][change_source] = new_nodes[:source_changes]
```

Comment:
Could this be replaced with:

```suggestion
            s_iloc[positive_count:][:(negative_count // 2)] = new_nodes[:(negative_count // 2)]
```

And vice versa with the line below? This would avoid computing `change_source` which is roughly 30% of the time in this function (based on some rough timing) and should still be random enough

Reply:
I'm nervous about doing this optimisation, without having any code/notebooks with which to validate that it's not degrading the models. I'll add a note in the code but not do it yet. How does that sound?

Reply:
Ah, also, I found that `np.random(size=...) < 0.5` is about 2&times; faster than `np.randint(2, size=...)`.

Reply:
That's a fair point, I think that sounds reasonable.